core/templates/services/prevent-reload-event.service.ts

```diff
@@ -0,0 +1,64 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Factory for giving warning before leaving page.
+ */
+
+import { Injectable } from '@angular/core';
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { WindowRef } from 'services/contextual/window-ref.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class PreventReloadEvent {
+  listenerActive: boolean;
+  constructor(private windowRef: WindowRef) {
+    this.listenerActive = false;
+  }
+
+  addListener(): void {
+    if (this.listenerActive) {
+      return;
+    }
+    this.windowRef.nativeWindow.addEventListener(
+      'beforeunload', this._confirmAlert);
+    this.listenerActive = true;
+  }
+
+  removeListener(): void {
+    this.windowRef.nativeWindow.removeEventListener(
+      'beforeunload', this._confirmAlert);
+    this.listenerActive = false;
+  }
+
+  private _confirmAlert(
+    e: { preventDefault: () => void; returnValue: string; }
+  ): void {
+    // HTML specification states that authors should use the
+    // below function.
+    e.preventDefault();
+    // The exact value in returnValue is not relevant, but it needs to be set
+    // in order to trigger a confirmation before leaving.
+    e.returnValue = '';
+  }
+
+  ngOnDestroy(): void {
```

Comment:
Do we need ngOnDestroy for service?

Reply:
Hi @DubeySandeep, services also have the `ngOnDestroy` lifecycle hook (see [docmentation](https://angular.io/api/core/OnDestroy#methods)). When the component using this service is destroyed, the service also gets destroyed.